src/components/waveform-canvas.js

```diff
@@ -0,0 +1,172 @@
+import PropTypes from 'prop-types';
+import React, { useEffect, useRef } from 'react';
+import { IMAGE_OPTIMIZATION_THRESHOLD_WIDTH } from '../domain/constants.js';
+
+const DEFAULT_OPTIONS = {
+  width: IMAGE_OPTIMIZATION_THRESHOLD_WIDTH,
+  height: Math.round(IMAGE_OPTIMIZATION_THRESHOLD_WIDTH / 2.5),
+  penWidth: 2,
+  smoothing: true,
+  penColor: '#666666',
+  baselineColor: '#333333',
+  backgroundColor: '#eeeeee'
+};
+
+class Wavedrawer {
+  constructor(element, options = {}) {
+    const canvas = element.ownerDocument.createElement('canvas');
+    canvas.setAttribute('style', 'display: block;');
+    element.replaceChildren(canvas);
+
+    this._canvas = canvas;
+    this._ctx = canvas.getContext('2d');
+    this._options = { ...DEFAULT_OPTIONS, ...options };
+    this._values = Array.from({ length: this._options.width }, () => 0);
+
+    this._initCanvas();
+    this._draw();
+  }
+
+  setOptions(options) {
+    const newOptions = { ...this._options, ...options };
+    const renderCriticalOptions = ['width', 'height', 'penColor', 'baselineColor', 'backgroundColor'];
+    const needsRerendering = renderCriticalOptions.some(key => newOptions[key] !== this._options[key]);
+
+    this._options = newOptions;
+
+    if (this.width !== this._values.length) {
+      // Adjust values array size to changed width
+      const oldValues = this._values;
+      const factor = oldValues.length / this._options.width;
+      const mapIndex = index => Math.max(0, Math.min(oldValues.length - 1, Math.round(index * factor)));
+      this._values = Array.from({ length: this._options.width }, (_, index) => oldValues[mapIndex(index)]);
+    }
+
+    if (needsRerendering) {
+      this._canvas.width = this._options.width;
+      this._canvas.height = this._options.height;
+      this._draw();
+    }
+  }
+
+  clear() {
+    this._values = Array.from({ length: this._options.width }, () => 0);
+    this._draw();
+  }
+
+  destroy() {
+    this._canvas.onmousedown = null;
+    this._canvas.onmousemove = null;
+    this._canvas.remove();
+    this._canvas = null;
+    this._ctx = null;
+  }
+
+  _initCanvas() {
+    this._canvas.width = this._options.width;
+    this._canvas.height = this._options.height;
+    this._canvas.onmousedown = event => {
+      if (event.button === 0) {
+        this._handleUserDraw(event.offsetX, event.offsetY);
+      }
+    };
+    this._canvas.onmousemove = event => {
+      if (event.buttons === 1) {
```

Comment:
Above you use `event.button` and here `event.buttons`, I presume this is not correct 🤔 
What are `0` and `1` here, by the way? Could we put them in named consts maybe to explain?

Reply:
It's correct. The `button` property indicates the button involved on events such as `mousedown`/`mouseup`, the `buttons` property indicates which of the buttons are pressed during non-button-related events such as `mousemove`. Even worse, their enum values are different. On `button`, `0` means main button, on `buttons`, `1` means main button. I extracted those as constants as you suggested.

Reply:
Pfu, anarchy. Thanks for making it clearer!